src/Interpreters/InterpreterSystemQuery.cpp

```diff
@@ -1219,26 +1229,193 @@ void InterpreterSystemQuery::dropReplica(ASTSystemQuery & query)
+
+        table_name_to_metadata.emplace(unescapeForFileName(escaped_table_names[i]), std::move(res.data));
+    }
+    String create_query = fmt::format("CREATE DATABASE `{}` ENGINE=Atomic", restoring_database_name);
+
+    auto create_ctx = Context::createCopy(Context::getGlobalContextInstance());
+    executeQuery(create_query, create_ctx, QueryFlags{.internal = true});
+
+    TablesDependencyGraph tables_dependencies("Memory (" + restoring_database_name + ")");
+    LOG_TEST(log, "table_name_to_metadata size={}", table_name_to_metadata.size());
+
+    for (const auto & [table_name, create_table_query] : table_name_to_metadata)
+    {
+        LOG_TEST(log, "table_name_to_metadata: table_name={}", table_name);
+        /// Note that table_name could contain a dot inside (e.g. .inner.1234-1234-1234-1234)
+        /// And QualifiedTableName::parseFromString doesn't handle this.
+        auto qualified_name = QualifiedTableName{.database = restoring_database_name, .table = table_name};
+        auto query_ast = DatabaseReplicated::parseQueryFromMetadataInZooKeeper(
+            getContext(),
+            /*database_name_=*/restoring_database_name,
+            /*zookeeper_path_=*/zookeeper_path,
+            /*node_name=*/table_name,
+            /*query=*/create_table_query);
+        tables_dependencies.addDependencies(
+            qualified_name,
+            getDependenciesFromCreateQuery(getContext()->getGlobalContext(), qualified_name, query_ast, getContext()->getCurrentDatabase())
+                .dependencies);
+    }
+
+    auto tables_to_create_by_level = tables_dependencies.getTablesSplitByDependencyLevel();
+    for (const auto & tables_to_create : tables_to_create_by_level)
+    {
+        for (const auto & storage_id : tables_to_create)
+        {
+            auto table_name = storage_id.getTableName();
+
+            auto metadata_it = table_name_to_metadata.find(table_name);
+            if (metadata_it == table_name_to_metadata.end())
+            {
+                /// getTablesSortedByDependency() may return some not existing tables or tables from other databases
+                LOG_WARNING(
+                    log,
+                    "Got table name {} when resolving table dependencies, "
+                    "but database {} does not have metadata for that table. Ignoring it",
+                    storage_id.getNameForLogs(),
+                    restoring_database_name);
+                continue;
+            }
+            const auto & create_query_string = metadata_it->second;
+            auto query_ast = DatabaseReplicated::parseQueryFromMetadataInZooKeeper(
+                getContext(),
+                /*database_name_=*/restoring_database_name,
+                /*zookeeper_path_=*/zookeeper_path,
+                /*node_name=*/table_name,
+                /*query=*/create_query_string);
+            auto create_query_context = Context::createCopy(getContext());
```

Comment:
We need to enable all the experimental settings, and a lot of other things, similar to `recoverLostReplica` in `DatabaseReplicated`:

https://github.com/ClickHouse/ClickHouse/blob/ced5264967018bc2fdfdcc5edf6cbc8be7ae89c2/src/Databases/DatabaseReplicated.cpp#L1369-L1395

I would also probably to it with `LoadingStrictnessLevel::FORCE_RESTORE`.